applications/FluidDynamicsApplication/custom_processes/compute_one_side_nodal_pressure_gradient_process.cpp

```diff
@@ -0,0 +1,148 @@
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Mohammad R. Hashemi
+//
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/parallel_utilities.h"
+
+// Application includes
+#include "fluid_dynamics_application_variables.h"
+#include "compute_one_side_nodal_pressure_gradient_process.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+
+/// constructor
+ComputeOneSideNodalPressureGradientProcess::ComputeOneSideNodalPressureGradientProcess(
+        ModelPart& rModelPart)
+    : Process(), mrModelPart(rModelPart) {}
+
+/// Constructor with Kratos parameters.
+ComputeOneSideNodalPressureGradientProcess::ComputeOneSideNodalPressureGradientProcess(
+    ModelPart& rModelPart,
+    Parameters Parameters)
+    : ComputeOneSideNodalPressureGradientProcess(
+        rModelPart
+    ){}
+
+/// Constructor with Kratos model
+ComputeOneSideNodalPressureGradientProcess::ComputeOneSideNodalPressureGradientProcess(
+    Model& rModel,
+    Parameters Parameters)
+    : ComputeOneSideNodalPressureGradientProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString())
+    ){}
+
+void ComputeOneSideNodalPressureGradientProcess::Execute(){
+
+    KRATOS_TRY;
+
+    // Set to zero
+    block_for_each(mrModelPart.Nodes(), [&](Node<3>& rNode){
+        rNode.SetValue(NODAL_AREA, 0.0);
+        rNode.FastGetSolutionStepValue(PRESSURE_GRADIENT).clear();
+    });
+
+    // Current domain size
+    const unsigned int num_dim = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+    const unsigned int num_nodes = num_dim + 1; //For tetrahedra and triangles
+
+    // Auxiliar containers
+    double detJ0 = 0.0;
+    Matrix DN_DX, J0, InvJ0;
+    Vector N(num_nodes), pressures(num_nodes), distances(num_nodes), grad(num_dim);
+
+    // First element iterator
+    const auto it_element_begin = mrModelPart.ElementsBegin();
+
+    // Iterate over the elements
+    #pragma omp parallel for firstprivate(DN_DX, N, J0, InvJ0, detJ0, pressures, distances, grad)
+    for(int i_elem=0; i_elem<static_cast<int>(mrModelPart.Elements().size()); ++i_elem) {
+
+        auto it_elem = it_element_begin + i_elem;
+        auto& r_geometry = it_elem->GetGeometry();
+
+        // Current geometry information
+        const unsigned number_of_nodes = r_geometry.PointsNumber();
+
+        for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node){
+            distances(i_node) = r_geometry[i_node].FastGetSolutionStepValue(DISTANCE);
+        }
+
+        unsigned int nneg=0, npos=0;
+        for(unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            if(distances(i) > 0) npos += 1;
+            else /* if(distances(i) < 0) */ nneg += 1;
+        }
+
+        if(nneg == 0 || npos == 0)
```

Comment:
This means that the pressure gradient is computed for all the non-intersected elements (both sides of the level set). If I'm not wrong, the class name is then inconsistent with the process operations.

Reply:
It actually separates the pressure field and computes pressure gradient specifically for each side, i.e. liquid pressure only contributes to the liquid pressure gradient calculated on negative nodes. Similarly, the gas pressure gradient is calculated on positive nodes.

Reply:
Yes, but by one side what comes to my mind is that the gradients are computed in either the air or the fluid side, but not in both sides (despite the gradients are computed with the corresponding side values).

Reply:
What about something like:
compute_**separated**_nodal_pressure_gradient_process

Reply:
And `calculate_consistent_nodal_pressure_gradient_process`?

Indeed, the same procedure can be templated for any variable type so we can use it in any levelset-based formulation. Having  said this, we can rename it to `calulate_levelset_consistent_gradient_process` (note that we normally use the keyword `calculate_` for this type of processes).

In a future PR we can extend the capabilities of the process to any variable value.